torchsde/_core/methods/euler_heun.py

```diff
@@ -0,0 +1,45 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .. import base_solver
+from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS
+
+
+class EulerHeun(base_solver.BaseSDESolver):
+    weak_order = 1.0
+    sde_type = SDE_TYPES.stratonovich
+    noise_types = NOISE_TYPES.all()
+    levy_area_approximations = LEVY_AREA_APPROXIMATIONS.all()
+
+    def __init__(self, sde, **kwargs):
+        if sde.noise_type == NOISE_TYPES.additive:
+            self.strong_order = 1.0
+        else:
+            self.strong_order = 1.0
```

Comment:
The above if statement is superfluous. Also, the strong order should be 0.5 for general noise.

Reply:
Ah, sorry for that, fixed. 
Also I've started working on a benchmark from [19] from the paper ("Smoking adjoints"). For now I've tried to follow exactly their formulas and they use Euler solver in log form - I need to compute `g` there but for `adjoint_sdeint` it fails as it's not supported due to efficiency. What can I do about it? Can `g` somehow be computed for `adjoint_sdeint`? (I've also tried to make it as a field in sde and access it from `_base_sde` from solver but as in forward it's wrapped: `ForwardSDE(base_sde)` in backward I think it's `AdjointSDE(ForwardSDE(base_sde))`)

Reply:
So back when that paper was written, "adjoint" meant "backprop through the solver".
These days, confusingly, we've appropriated the term to mean "don't backprop through the solver, and solve a differential equation instead".
As a result that paper is basically just about the now well-understood advantages of using backpropgation over forward-mode differentiation. (At least in this context.)

So they don't actually use the adjoint SDE at all, and what they do is equivalent to just using our `sdeint` in the normal way.

Reply:
Yes, I got it while reading the paper. What I'm trying is to extend that benchmark from "smoking adjoints" with stochastic adjoint from here so I want to do it `forward` vs `sdeint` vs `adjoint_sdeint`. 
And due to the fact that in custom log_euler solver that I created I'm using `g` the described error occurs.

Reply:
I'm not sure what your distinction between `forward` and `sdeint` is.

You shouldn't need to create a custom solver I don't think - IIRC in that paper they use an SDE to model log-rates; solve that with Euler--Maruyama; take an exponential if you want the rates themselves.

Likewise, you shouldn't need to use `g` in the adjoint: the standard EM scheme only needs `g_prod`.

Does that make sense / am I understanding you correctly? If not perhaps can you provide some references to where in the SA paper you're looking? (There's a couple copies of that paper out there, I tend to look at this one - https://www0.gsb.columbia.edu/faculty/pglasserman/Other/RiskJan2006.pdf)

Reply:
Yeah I was reading exactly that one. 

By `forward` method I mean computing `dV/dS` with `(sdeint(S+h) - sdeint(S-h)) / 2h`, by `sdeint` ("adjoint" in "smoking adjoints", "backprop through the solver" here 😄) `dV/dS` is `sol = sdeint(); torch.grad(sol, S)` and last one `adjoint_sdeint` is `dV/dS` is `sol = adjoint_sdeint(); torch.grad(sol, S)`. (I hope I didn't mess it up). Also I've got PoC and they yield same gradient result.

Code is similar to this (as `forward` vs `sdeint`): https://github.com/mgroncki/AdjointMonteCarlo/blob/master/PlainVanillaOption/blackMC.cpp

Thanks for help! I will try it that way. To explain myself the log-Euler solver was used in paper and also in references in `Monte Carlo Methods In Financial Engineering` (page 398) directly for that model.

Reply:
But I think I need a bit more time to fully figure all these methods out on paper. 
The chapter 7. of that book also makes comparison of two methods: the finite-difference approach defined as `(Y(S+h) - Y(S-h)) / 2h` and `pathwise` mehtod.